### ISSUE TITLE
Temporal utilities miss variations due to daylight savings

### ISSUE THREAD

**Issue by novemberkilo** [apache/arrow-rs#901]
```

    #[cfg(feature = "chrono-tz")]
    #[test]
    fn test_temporal_array_timestamp_hour_with_dst_timezone_using_chrono_tz() {
        //
        // 1635577147 converts to 2021-10-30 17:59:07 in time zone Australia/Sydney (AEDT)
        // The offset (difference to UTC) is +11:00. Note that daylight savings is in effect on 2021-10-30.
        // When daylight savings is not in effect, Australia/Sydney has an offset difference of +10:00.

        let a = TimestampMillisecondArray::from_opt_vec(
            vec![Some(1635577147000)],
            Some("Australia/Sydney".to_string()),
        );
        let b = hour(&a).unwrap();
        assert_eq!(17, b.value(0));
    }

```

This test [fails](https://github.com/novemberkilo/arrow-rs/commit/473e2e4eb4f3b47aa7110d1a23444647b1109665) - it returns 16 instead of 17 and demonstrates that the offset for the timezone is not correctly taking daylight savings into account.

__Originally posted by @novemberkilo in https://github.com/apache/arrow-rs/pull/849#issuecomment-955647565__

**Comment by novemberkilo** [comment]
I have a fix in the works - please assign to me as appropriate. // @alamb

**Comment by novemberkilo** [comment]
@alamb I have a fix on a branch that builds off my work in #849 -- here is the diff https://github.com/novemberkilo/arrow-rs/compare/master...novemberkilo:topic/dst-bug-temporal

I suggest we work through #849 and then I will put up a PR separately.
